snapping control

mprins opened the pull request:
This is ready for review 2.

actions:
- [x] add a OL only instruction to components.json
- [x] remove console logging
- [x] refactor the snapping controller
  - [x] move class from viewer.components.OpenLayersSnappingController to viewer.viewercontroller.openlayers.OpenLayersSnappingController
  - [x] provide (abstract) accessor in viewerController or use `createComponent`
  - [x] add factory method in MapComponent to get a SnappingController
- [x] remove unused method `layerRemoved` from snapping controller
- [x] rebase on upstream/master
- [x] fix the sprite (027b05683338575e3c902a02a019e88194ec99d0 and 7c2209118231754b857956ec3df12eb52b8a97c7) the sprite now also has split and merge buttons

mprins committed:
snapping control + config control

mprins committed:
don't ignore the filter options if explicitly specified

mprins committed:
voeg sprite toe voor knop en popover

mprins committed:
fix trailing comma

mprins committed:
(admin) help voor de snapping control

mprins committed:
(admin) help voor de snapping control

mtoonen commented on viewer/src/main/java/nl/b3p/viewer/stripes/AttributesActionBean.java:
```diff
@@ -97,6 +97,9 @@
     private List<String> attributesToInclude = new ArrayList();
 
+    /**
+     * To force returning geometry set this to true in the request.
```
I think this only works when you set the geometry in attributesToInclude.
I think it's better to change the variablename to reflect it's purpose.

mtoonen commented on viewer/src/main/webapp/viewer-html/components/OpenLayersSnappingController.js:
```diff
@@ -0,0 +1,259 @@
+     * @param {type} map
+     * @param {type} options
+     */
+    layerRemoved: function (map, options) {
```
Doesn't do much anymore. Remove function and listener?

mtoonen commented on viewer/src/main/webapp/viewer-html/components/OpenLayersSnappingController.js:
```diff
@@ -0,0 +1,259 @@
+            return;
+        }
+        // filter for edit and drawing control layers
+        if ((Ext.String.startsWith(options.layer.name, "drawing", true)) ||
```
Seems fragile to depend on the layername, since it can change. Don't have any better ideas though..

mtoonen commented on viewer/src/main/webapp/viewer-html/components/OpenLayersSnappingController.js:
```diff
@@ -0,0 +1,259 @@
+    deactivate: function () {
+        this.frameworkControl.deactivate();
+    },
+    parseFeatures: function (data) {
```
Is it fast enough? Can imagine when having large datasets (or a zoomed out extent), a lot of features are being retrieved and shown.
Client side caching?

mtoonen commented on viewer/src/main/webapp/viewer-html/components/OpenLayersSnappingController.js:
```diff
@@ -0,0 +1,259 @@
+     */
+    constructor: function (config) {
+        viewer.components.OpenLayersSnappingController.superclass.constructor.call(this, config);
+        this.frameworkMap = this.config.viewerController.mapComponent.getMap().getFrameworkMap();
```
Overall note: 
It's too openlayers specific. I think it has to be made more framework independent, but you're already there.
The instantiation should be in OpenLayersMapComponent or OpenLayersVectorLayer (or somewhere else, but in a place which knows what framework is used).
I think it would also be good to make a superclass for the snappingcontroller, which has the basic implementation (like adding features to the vectorlayers, listening to the events, etc.), so the framework specific implementation should be quite limited and restricted to framework specific calls (and probably more of a passthrough for calls made by Snapping.js).
Framework specific files should be under it's framework folder for clarity.

mtoonen commented on viewer/src/main/webapp/viewer-html/components/Snapping-config.js:
```diff
@@ -0,0 +1,59 @@
+        }
+        configObject.showLabelconfig = true;
+        viewer.components.CustomConfiguration.superclass.constructor.call(this, parentId, configObject);
+        this.createCheckBoxes(this.configObject.layers, {
```
Bij het maken van een kopie veranderen de layer ids. Bij bijvoorbeeld Edit, Attributelist en transparency zie je hoe deze omgezet worden als dit gebeurd.

mprins replied:
dit werkt al uit de doos

mtoonen commented on viewer/src/main/webapp/viewer-html/components/Snapping.js:
```diff
@@ -0,0 +1,204 @@
+
+            if (item.checked) {
+                if (idx < 0) {
+                    console.debug("add data for layer", item.inputValue, appLayer.layerName);
```
console weghalen

mtoonen commented on viewer/src/main/webapp/viewer-html/components/Snapping.js:
```diff
@@ -0,0 +1,204 @@
+                }
+            } else {
+                if (idx > -1) {
+                    console.debug("remove data for layer", item.inputValue, appLayer.layerName);
```
console weghalen

mtoonen commented on viewer/src/main/webapp/viewer-html/components/Snapping.js:
```diff
@@ -0,0 +1,204 @@
+        });
+    },
+    createController: function () {
+        this.snapCtl = Ext.create('viewer.components.OpenLayersSnappingController', {
```
Zoals eerder gezegd is dit niet de plek waar dit geinstantieerd moet worden (in ieder geval niet de framework specifieke controller)

mtoonen commented on viewer/src/main/webapp/viewer-html/components/components.json:
```diff
@@ -672,6 +672,18 @@
+        name: "Snapping",
+        shortName: "Sn",
+        group: "Component",
+        sources: ["OpenLayersSnappingController.js", "Snapping.js"],
```
nog toevoegen dat het (voor nu) alleen openlayers is:
        notInCombinationWith: ["viewer.mapcomponents.FlamingoMap"]

mprins committed:
add notInCombinationWith: ["viewer.mapcomponents.FlamingoMap"]

mprins committed:
remove console logging

and also listen/attach to layers of the split control

mprins committed:
Remove unused method `layerRemoved` from snapping controller


it seems controls like edit do not actually remove their editable layer from the map so it seems useless anyway

mprins committed:
refactor snapping controller

for better abstraction from openlayers framework

mprins committed:
fix layerswitching behaviour


don'y turn off (wms) layers from the snapping control that were switched on before we used them as a snapping target

mprins committed:
add icons to the sprite for split and merge

Conflicts:
	viewer/src/main/webapp/viewer-html/components/Component.js

mprins committed:
fix the tool icon positions of the sprite



Which were messed up since 36235b57acdacb0b978b73fe7a6d8e81e64ccd36

mprins committed:
Address some review issues in OpenLayersSnappingController 

  - use real error function in load attributes
  - only retrieve appLayer attributes once and if needed

mprins committed:
controls should register for snapping client themselves


Instead of looking at the (hardcoded) layername make the controls that wish to benefit from snapping (Edit, Drawing, SpatialFilter, ...) register with the viewerController when they create their editable layer and before they add that to the map.

Also improve handling of layerswitching in the control.
